Add test for opaque traffic between meshed and unmeshed workloads

mateiidavid opened the pull request:
### What

Fixes #5838 

* Opaque ports should still work on an unmeshed service. Currently, there
is no integration test to validate that traffic between a meshed app and
an unmeshed (but opaque) service works as expected.

* This change adds a new integration test that confirms opaque ports
behaves correctly in the given scenario. Correct behavior in this case
means that there will only be TCP metrics available on the inbound side
(and not on the outbound, unmeshed side). Furthermore, no HTTP metrics
should be recorded by either side (as consequence of opaqueness).

### How

I tried to keep as much as possible to the current style of the tests. There are three main implications that arise from opaque ports between a meshed and unmeshed workload:
  1. We have no L7 protocol metrics (inbound/outbound);
  2. No TCP metrics on server, TCP metrics on client (inbound yes, outbound no -- lack of proxy);
  3. No mTLS (unmeshed pod doesn't have the proxy TLS identity).
 
Based on these 3 implications I designed the test in the same format as the existing ones.
* First, we add an unmeshed service and pod -- I put these in a separate yaml file. The Service has an opaque ports annotation. I wanted to avoid reading the file contents in so I created a test helper function to just `kubectl apply -f` them 
(**Note**: *wanted to use an existing function but the ones we have in only execute commands with data from stdin, e.g `kubectl apply -f - `*).
We also add in a meshed SC that sends stuff over http to our unmeshed pod. We check the pods have spun up.
* Second, we check the inbound side. For the meshed app, we expect to see TCP metrics but no HTTP metrics. I added new regex expressions to account for no TLS. If either of these two conditions (TCP/HTTP) are not met, we fail.
* Third and final, we check the unmeshed service. If we have _any_ metrics, we fail. Our pod is unmeshed and the traffic is opaque, we wouldn't expect to have anything in there. I created an http response metric to make sure we don't get anything on the inbound.


I created the regex expressions based on what metrics look like when no TLS identity is discovered, I've appended them below in case someone wants to double check.
<details> <summary> Click me for metric examples </summary>

```sh
# TCP  tcp_open_connections{direction="inbound",peer="src",target_addr="0.0.0.0:4191",tls="no_identity",no_tls_reason="no_tls_from_remote"} 1      tcp_open_connections{direction="outbound",peer="src",target_addr="10.43.251.103:8080",tls="no_identity",no_tls_reason="loopback"} 1

# HTTP INBOUND (meshed pod)
response_latency_ms_bucket{direction="inbound",target_addr="0.0.0.0:4191",tls="no_identity",no_tls_reason="no_tls_from_remote",status_code="200",le="30000"}

# HTTP OUTBOUND (unmeshed pod)
request_total{direction="outbound",authority="svc-opaque-unmeshed.linkerd-opaque-ports-test.svc.cluster.local:8080",target_addr="10.42.0.169:8080",tls="no_identity",no_tls_reason="not_provided_by_service_discovery",dst_deployment="opaque-unmeshed-svc",dst_namespace="linkerd-opaque-ports-test",dst_pod="opaque-unmeshed-svc-5f9bd44cfc-thcpm",dst_pod_template_hash="5f9bd44cfc",dst_service="svc-opaque-unmeshed",dst_serviceaccount="default"} 16
```
</details>

Let me know if I missed anything or if I can improve this. Naming suggestions also very welcome, feel like I didn't do a great job on the naming side this time around 😄. 

Signed-off-by: Matei David <matei@buoyant.io>

kleimkuhler reviewed (commented):
I have a suggestion for keeping this one testdata file which should remove some of these changes needed right now.

Also, there will be no inbound metrics on the unmeshed service because `getPodMetrics` is just a `linkerd diagnostics proxy-metrics` under the hood; if there is no proxy then these metrics would be empty. Checking that they don't match certain regexes will always be true because it's going to be an empty string.

Therefore, I think the tests we have for the meshed client is what we'll need to go off of, and the existing checks I think are good enough.

mateiidavid commented on testutil/kubernetes_helper.go:
```diff
@@ -133,6 +133,17 @@ func (h *KubernetesHelper) KubectlApply(stdin string, namespace string) (string,
 	return h.Kubectl(stdin, args...)
 }
 
+// KubectlApplyFile applies a given configuration file in a namespace without
+// reading the file contents into memory. If no namespace is provided, it does
+// not specify the '--namespace' flag.
+func (h *KubernetesHelper) KubectlApplyFile(path string, namespace string) (string, error) {
+	args := []string{"apply", "-f", path}
+	if namespace != "" {
+		args = append(args, "--namespace", namespace)
+	}
+	return h.Kubectl("", args...)
+}
+
```
Great suggestion. I completely missed we already have a `ReadFile` function to deal with this. I'll keep in the back of my head for the next time this comes up. To resolve this, I'll take your other suggestion posted below (annotate deployment with `inject: disabled`), that should eliminate the need to read the file in.

Also, will remove this method. Thanks!

mateiidavid commented:
Thanks @kleimkuhler for the review! I have addressed all of your comments:

* instead of checking in a new test file I took the manifests and added them to the existing one (with an `inject: disabled` annotation on the deployment);
* removed `ApplyWithFile` command (and its usage) in favour of `linkerd inject` -- since we have `inject: disabled`, running the entire file through `linkerd inject` won't actually add the sidecar to the unmeshed deployment;
* removed the unnecessary test and the HTTP response regex (used only to test the unmeshed workload).

Additionally, following an off-Github conversation, I removed the `--manual` flag from `LinkerdRun`. This shouldn't impact the test success. If it does, I'll add it back in.

kleimkuhler commented:
@mateiidavid Great! Looking at this with those changes, it probably does make sense to check a metric making sure that we do have _some_ outbound connection—especially since we can't rely off of outbound HTTP requests. Do you mind adding a metric check for open outbound TCP connections?

mateiidavid commented:
> @mateiidavid Great! Looking at this with those changes, it probably does make sense to check a metric making sure that we do have _some_ outbound connection—especially since we can't rely off of outbound HTTP requests. Do you mind adding a metric check for open outbound TCP connections?

Great suggestion! The test didn't feel very thorough to me once I realised we can't check for outbound HTTP requests. I like the idea of testing for an outbound connection.

I pushed the changes; since outbound TCP connections to unmeshed workloads don't have any TLS I had to explicitly add that in the regex pattern -- I had hoped at first I could come up with a more generic regex pattern for outbound TCP connections but I thought in the end having a pattern for this case specifically (meshed to unmeshed workload) would be better.

wdyt @kleimkuhler, let me know if you have any other suggestions.

kleimkuhler reviewed (commented):
The last thing to do here is remove the test for the inbound TCP metric. The unmeshed application shouldn't be making any requests _to_ the slow cooker, so the fact that the slow cooker does have inbound connections means it's probably coming from k8s or something like that. It's only the outbound that we care about—that there are no outbound HTTP requests and it has at least some outbound TCP connection open.

kleimkuhler reviewed (approved):
Awesome thanks @mateiidavid!

Pothulapati commented on test/integration/opaqueports/opaque_ports_test.go:
```diff
@@ -15,16 +15,24 @@ import (
 		`request_total\{direction="outbound",authority="[a-zA-Z\-]+\.[a-zA-Z\-]+\.svc\.cluster\.local:8080",target_addr="[0-9\.]+:8080",tls="true",.*`,
 	)
+	httpRequestTotalUnmeshedRE = regexp.MustCompile(
+		`request_total\{direction="outbound",authority="[a-zA-Z\-]+\.[a-zA-Z\-]+\.svc\.cluster\.local:8080",target_addr="[0-9\.]+:8080",tls="no_identity",.*`,
```
Why not check for the specific authority i.e `svc-opaque-unmeshed.linkerd-opaque-ports-test.svc.cluster.local:8080` as we know it wil lbe that for sure?

mateiidavid replied:
No reason in particular, think at the time it felt more intuitive to me to keep it as any authority in case we might want to reuse it later. Since I doubt this will be the case in the near future I'll change it to be explicit 👍🏻

Pothulapati replied:
Makes sense!

Pothulapati reviewed (approved):
LGTM overall, except for the nit.
